src/subprotocols/sumcheck.rs

```diff
@@ -25,6 +25,133 @@ use crate::msm::VariableBaseMSM;
 #[cfg(feature = "multicore")]
 use rayon::prelude::*;
 
+#[derive(Clone, PartialEq)]
+pub enum CubicSumcheckType {
+  Prod,
+  Flags
+}
+
+pub struct CubicSumcheckParams<F: PrimeField> {
+  poly_As: Vec<DensePolynomial<F>>,
+  poly_Bs: Vec<DensePolynomial<F>>,
+
+  a_to_b: Vec<usize>,
+
+  poly_eq: DensePolynomial<F>,
+
+  pub num_rounds: usize,
+
+  pub sumcheck_type: CubicSumcheckType 
+}
+
+impl<F: PrimeField> CubicSumcheckParams<F> {
+  pub fn new_prod(
+    poly_lefts: Vec<DensePolynomial<F>>, 
+    poly_rights: Vec<DensePolynomial<F>>, 
+    poly_eq: DensePolynomial<F>, 
+    num_rounds: usize) -> Self {
+      debug_assert_eq!(poly_lefts.len(), poly_rights.len());
+      debug_assert_eq!(poly_lefts[0].len(), poly_rights[0].len());
+      debug_assert_eq!(poly_lefts[0].len(), poly_eq.len());
+
+      let a_to_b = (0..poly_lefts.len()).map(|i| i).collect();
+
+      CubicSumcheckParams {
+        poly_As: poly_lefts,
+        poly_Bs: poly_rights,
+        a_to_b,
+        poly_eq,
+        num_rounds,
+        sumcheck_type: CubicSumcheckType::Prod
+      }
+  }
+
+  /// flag_map: poly_leaves length vector mapping between poly_leaves indices and flag indices.
+  pub fn new_flags(
+    poly_leaves: Vec<DensePolynomial<F>>, 
+    poly_flags: Vec<DensePolynomial<F>>, 
+    poly_eq: DensePolynomial<F>, 
+    flag_map: Vec<usize>,
+    num_rounds: usize) -> Self {
+      debug_assert_eq!(poly_leaves.len(), flag_map.len());
+      debug_assert_eq!(poly_leaves[0].len(), poly_flags[0].len());
+      debug_assert_eq!(poly_leaves[0].len(), poly_eq.len());
+
+      CubicSumcheckParams { 
+        poly_As: poly_leaves, 
+        poly_Bs: poly_flags, 
+        a_to_b: flag_map, 
+        poly_eq, 
+        num_rounds, 
+        sumcheck_type: CubicSumcheckType::Flags 
+      }
+  }
+
+  // TODO(sragss): Due to inlining this may be more efficient with a lambda. Test with benchmarking.
+  pub fn combine(&self, a: &F, b: &F, c: &F) -> F {
+    match self.sumcheck_type {
+      CubicSumcheckType::Prod => Self::combine_prod(a, b, c),
+      CubicSumcheckType::Flags => Self::combine_flags(a, b, c),
+      _ => panic!("Unimpl")
+    }
+  }
```

Comment:
this feels somewhat like an anti-pattern, maybe we should consider turning this `impl` into a trait

Reply:
I wrote this out a few different ways and disagree.

If we do a different struct for each CubicSumcheckType we end up sharing most of the logic and duplicating it. 

Alternatively we could do a static generic on the struct: 
```
pub struct CubicSumcheckParams<F: PrimeField, S: CubicSumcheckType> {
    // ...
    _marker: PhantomData<S>
}

impl<F: PrimeField> CubicSumcheckParams<F> {
    // ...
    fn combine(a: &F, b: &F, c: &F) -> F {
        S::combine(a,b,c)
    }
}

trait CubicSumcheckType {
    fn combine(a: &F, b: &F, c: &F) -> F;
}
struct CubicSumcheckProd {}
impl CubicSumcheckType for CubicSumcheckProd {
    fn combine(a: &F, b: &F, c: &F) -> F {
        *a * *b * *c
    }
}
struct CubicSumcheckFlags {}
impl CubicSumcheckType for CubicSumcheckFlags {
    fn combine(a: &F, b: &F, c: &F) -> F {
        *a * (*b * *c + (F::one() - *c)
    }
}
```

I was hoping it would be cleaner but I think it's clearest as we have it. Especially given we don't expect the number of these impls to grow dramatically. If so, I think we should make the adjustments described above.